core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -609,6 +616,27 @@ private class BufferReleasingInputStream(
   override def reset(): Unit = delegate.reset()
 }
 
+/**
+ * A listener to be called at the completion of the ShuffleBlockFetcherIterator
+ * @param data the ShuffleBlockFetcherIterator to process
+ */
+private class ShuffleFetchCompletionListener(var data: ShuffleBlockFetcherIterator)
+  extends TaskCompletionListener {
+
+  override def onTaskCompletion(context: TaskContext): Unit = {
```

Comment:
There are two entry points to this, one is when iterator ends iteration, one is when task completion. Can they happen closely and both run into cleanup?

Reply:
No, that can never happen for they are running in the same task thread, so there are no thread safe problems, and the iterator ends callback will definitely run before task completion callback if there are no exceptions.